portal-kernel/src/com/liferay/portal/kernel/portlet/render/PortletRenderUtil.java

```diff
@@ -0,0 +1,592 @@
+					contextPath = PortalUtil.getPathContext();
+				}
+				else {
+					contextPath =
+						PortalUtil.getPathProxy() + portlet.getContextPath();
+				}
+
+				Collection<String> portletResources =
+					portletResourceAccessor.get(portlet);
+
+				for (String portletResource : portletResources) {
+					boolean module = false;
+
+					if (portletResource.startsWith("module:")) {
+						module = true;
+
+						portletResource = portletResource.substring(7);
+					}
+
+					if (!HttpComponentsUtil.hasProtocol(portletResource)) {
+						portletResource = PortalUtil.getStaticResourceURL(
+							httpServletRequest, contextPath + portletResource,
+							rootPortlet.getTimestamp());
+					}
+
+					if (!portletResource.contains(Http.PROTOCOL_DELIMITER)) {
+						String cdnBaseURL = themeDisplay.getCDNBaseURL();
+
+						portletResource = cdnBaseURL.concat(portletResource);
+					}
+
+					if (module) {
+						portletResource = "module:" + portletResource;
+					}
+
+					if (visitedURLs.contains(portletResource)) {
+						continue;
+					}
+
+					visitedURLs.add(portletResource);
+
+					urls.add(portletResource);
+				}
+			}
+		}
+
+		return urls;
+	}
+
+	private static Collection<String> _getURLs(
+		HttpServletRequest httpServletRequest,
+		Collection<PortletResourceAccessor> portletResourceAccessors,
+		Collection<Portlet> portlets, URLType urlType) {
+
+		boolean fastLoad;
+		Predicate<String> predicate = null;
+
+		ThemeDisplay themeDisplay =
+			(ThemeDisplay)httpServletRequest.getAttribute(
+				WebKeys.THEME_DISPLAY);
+
+		if (urlType == URLType.CSS) {
+			fastLoad = themeDisplay.isThemeCssFastLoad();
+		}
+		else if (urlType == URLType.JAVASCRIPT) {
+			fastLoad = themeDisplay.isThemeJsFastLoad();
+
+			predicate = resource -> !themeDisplay.isIncludedJs(resource);
+		}
+		else {
+			throw new UnsupportedOperationException(
+				"Unsupported URL type " + urlType);
+		}
+
+		Set<String> visitedURLs = (Set<String>)httpServletRequest.getAttribute(
+			WebKeys.PORTLET_RESOURCE_STATIC_URLS);
+
+		if (visitedURLs == null) {
+			visitedURLs = new LinkedHashSet<>();
+
+			httpServletRequest.setAttribute(
+				WebKeys.PORTLET_RESOURCE_STATIC_URLS, visitedURLs);
+		}
+
+		if (fastLoad) {
+			Theme theme = themeDisplay.getTheme();
+
+			return _getComboServletURLs(
+				portletResourceAccessors, portlets, predicate,
+				theme.getTimestamp(),
+				PortalUtil.getStaticResourceURL(
+					httpServletRequest,
+					themeDisplay.getCDNDynamicResourcesHost() +
+						themeDisplay.getPathContext() + "/combo",
+					"minifierType=&themeId=" + themeDisplay.getThemeId(), -1),
+				visitedURLs);
+		}
+
+		return _getStaticURLs(
+			httpServletRequest, portletResourceAccessors, portlets,
+			visitedURLs);
+	}
+
+	private static void _writeCSSPath(PrintWriter printWriter, String cssPath) {
+		printWriter.print("<link href=\"");
+		printWriter.print(HtmlUtil.escape(cssPath));
+		printWriter.println("\" rel=\"stylesheet\" type=\"text/css\" />");
```

Comment:
Don't we need `data-senna-track="temporary"` here?

Reply:
It was previously there, https://github.com/liferay-frontend/liferay-portal/pull/2580/files#diff-1afd0967f8255757d64743e17acb070730b3094f09a0aa78a9608b4e2b379595L31

Reply:
argh!!! :scream: And there's an ID too. I need to fix it :-(

Reply:
Fixed in a new commit